engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -104,46 +104,73 @@ public void setTailStatus(TailStatus tailStatus) {
   public abstract Object execute(
       VirtualFrame frame, State state, UnresolvedSymbol symbol, Object self, Object[] arguments);
 
-  @Specialization(guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
-  Object doFunctionalDispatch(
+  @Specialization(guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)", "cachedSymbol == symbol"})
+  Object doFunctionalDispatchCachedSymbol(
       VirtualFrame frame,
       State state,
       UnresolvedSymbol symbol,
       Object self,
       Object[] arguments,
       @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
-      @Cached MethodResolverNode methodResolverNode) {
+      @Cached MethodResolverNode methodResolverNode,
+      @Cached("symbol") UnresolvedSymbol cachedSymbol,
+      @Cached("anyFunction(cachedSymbol)") Function anyFun) {
 
     Type selfTpe = typesLibrary.getType(self);
-    Function function = methodResolverNode.expectNonNull(self, selfTpe, symbol);
+    Function function = methodResolverNode.expectNonNull(self, selfTpe, cachedSymbol);
 
     RootNode where = function.getCallTarget().getRootNode();
     // If both Any and the type where `function` is declared, define `symbol`
     // and the method is invoked statically, i.e. type of self is the eigentype,
     // then we want to disambiguate method resolution by always resolved to the one in Any.
     if (where instanceof MethodRootNode node && typeCanOverride(node, EnsoContext.get(this))) {
-      Function anyFun = symbol.getScope().lookupMethodDefinition(EnsoContext.get(this).getBuiltins().any(), symbol.getName());
       if (anyFun != null) {
         function = anyFun;
       }
     }
     return invokeFunctionNode.execute(function, frame, state, arguments);
   }
 
+  Function anyFunction(UnresolvedSymbol symbol) {
+    Type any = EnsoContext.get(this).getBuiltins().any();
+    return symbol.getScope().lookupMethodDefinition(any, symbol.getName());
+  }
+
   private boolean typeCanOverride(MethodRootNode node, EnsoContext ctx) {
     Type methodOwnerType = node.getType();
     Builtins builtins = ctx.getBuiltins();
     Type any = builtins.any();
     Type warning = builtins.warning();
     Type panic = builtins.panic();
     return methodOwnerType.isEigenType()
-
             && builtins.nothing() != methodOwnerType
             && any.getEigentype() != methodOwnerType
             && panic.getEigentype() != methodOwnerType
             && warning.getEigentype() != methodOwnerType;
   }
 
+  @Specialization(replaces = "doFunctionalDispatchCachedSymbol", guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
+  Object doFunctionalDispatchUncachedSymbol(
```

Comment:
Is this really a 1:1 copy of `doFunctionalDispatchCachedSymbol`, just without caching the `anyFunction` resolution? Why is this necessary? What is the goal?

For each `UnresolvedSymbol symbol` there is only one query to `anyFunction(symbol)` (in a single `EnsoContext`) needed. It is possible that `cachedSymbol != symbol` (certaily in multiple contexts), but how can that happen in a single context? Does that happen? When?

In any case, the solution for multiple symbols is to built a polymorphic cache of certain depth/limit - e.g. use `limit=".."` in the previous `@Specialization`.

Then there has to be a "fallback" case when the limit is exceeded. Is this `@Specialization` supposed to be a fallback?

Reply:
I guess that this was needed, because `ModuleScope.lookupMethodDefinition` is an expensive method?

> Is it possible that cachedSymbol != symbol?

Is that even possible in multiple context? I mean, Enso method calls are "statically resolvable", are they not? Which means that for every call site, there is a fixed symbol. Am I missing something?

Reply:
> I guess that this was needed, because `ModuleScope.lookupMethodDefinition` is an expensive method

Yes, it is pretty expensive especially on `Any`. The reason why Benchmarks were timing out.

> It is possible that `cachedSymbol != symbol`

I was getting a lot of failures on `Value_Type`'s tests without this condition. Also `invokeMember` wouldn't play nice, as indicated by failures in `TypeMembersTest`.

Reply:
I think I addressed the comments in the latest change. It's good feedback btw, thank you.

Reply:
`replaces = "doFunctionalDispatchCachedSymbol",` seems reasonable.